Hmusum/allow groups to be down

hmusum opened the pull request.

Commit:
Remove interface now that there is only one implementation

Commit:
Remove unused method and simplify

Commit:
Minor cleanup, no functional changes

Commit:
Fix bug in distributuon config generation

Commit:
Allow more than 1 group in a content to be down at the same time

Based on config, all functional changes guarded by config
field max_number_of_groups_allowed_to_be_down in fleetcontroller config

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -441,4 +500,25 @@ private Result checkDistributors(Node node, int clusterStateVersion) {
+                                      .map(NodeInfo::getGroup)
+                                      .filter(Objects::nonNull)
+                                      .filter(Group::isLeafGroup)
+                                      .map(Group::getIndex)
```
re: offline discussion about the many fun and exciting quirks of groups, consider `Group::getPath` to disambiguate nested groups with the same logical index

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -441,4 +500,25 @@ private Result checkDistributors(Node node, int clusterStateVersion) {
+        var groups = new HashMap<Integer, Group>();
+        for (NodeInfo nodeInfo : clusterInfo.getStorageNodeInfos()) {
+            Group group = nodeInfo.getGroup();
+            int index = group.getIndex();
```
This should probably also use `getPath` to be nesting-agnostic

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -354,7 +398,22 @@ private static boolean groupContainsNode(Group group, Node node) {
+
     private Result checkAllNodesAreUp(ClusterState clusterState) {
+        // TODO Revisit this, check that all nodes in other groups are up?
+        if (moreThanOneGroupAllowedToBeDown()) return allowSettingOfWantedState();
```
I think we need to ponder and discuss the desired semantics for this one a bit, probably when @hakonhall is back

hmusum replied:
Not sure If I understand how more than one node can be set to maintenance (as is the goal when allowing more than one group to be down at a time) if that is the case? This methods checks both wanted state and actual state, so name is a bit misleading.

hakonhall replied:
Discussed offline

vekterli reviewed (dismissed):
:+1: some edge case stuff we probably need to ponder a bit before rolling out the feature

@hakonhall please also have a look

Commit:
Restore isControlledShutdown and simplify tests a bit

hmusum commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -268,6 +289,29 @@ private Result anotherNodeInAnotherGroupHasWantedState(StorageNodeInfo nodeInfo)
+     * Returns a disallow-result if there is other groups with nodes that are allowed to be down
+     * so that the number of groups > maxNumberOfGroupsAllowedToBeDown
+     */
+    private Result nodesInOtherGroups(Node node) {
```
Yes, will be renamed

hmusum commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/StateChangeHandler.java:
```diff
@@ -362,10 +355,10 @@ private boolean reportDownIfOutdatedSlobrokNode(ClusterState currentClusterState
+    private boolean isNotControlledShutdown(NodeState state) {
+        return (state.getState() != STOPPING
+                || (!state.getDescription().contains("Received signal 15 (SIGTERM - Termination signal)")
+                && !state.getDescription().contains("controlled shutdown")));
```
I'll create a wrapper method instead

hmusum commented on clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeCheckerTest.java:
```diff
@@ -53,7 +54,11 @@ private static ClusterState clusterState(String state) {
+    }
+
+    private static ClusterState defaultAllUpClusterState(int nodeCount) {
+        return clusterState(String.format("version:%d distributor:" + nodeCount + " storage:" + nodeCount, currentClusterStateVersion));
```
WIll fix

hmusum commented on clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeCheckerTest.java:
```diff
@@ -159,12 +164,106 @@ void testSafeMaintenanceDisallowedWhenOtherStorageNodeInFlatClusterIsSuspended()
         ContentCluster cluster = createCluster(4);
-        cluster.clusterInfo().getDistributorNodeInfo(0)
-                .setWantedState(new NodeState(DISTRIBUTOR, DOWN).setDescription("Orchestrator"));
+        setDistributorNodeWantedState(cluster, 0, new NodeState(DISTRIBUTOR, DOWN), "Orchestrator");
```
Yup

hmusum commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -268,6 +289,29 @@ private Result anotherNodeInAnotherGroupHasWantedState(StorageNodeInfo nodeInfo)
+     * so that the number of groups > maxNumberOfGroupsAllowedToBeDown
+     */
+    private Result nodesInOtherGroups(Node node) {
+        log.log(INFO, maxNumberOfGroupsAllowedToBeDown + " groups allowed to be down");
```
Yes, just for now, will change it in a later commit in this PR

hmusum commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -441,4 +500,25 @@ private Result checkDistributors(Node node, int clusterStateVersion) {
+                                      .map(Group::getIndex)
+                                      .collect(Collectors.toSet());
+
+        // Groups are not equal even if they have the same index
```
`Group.equals()` considers many more things than just index, so it's not usable in this context. Will probably rewrite this part anyway.

hmusum commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -214,26 +224,37 @@ private Result canSetStateMaintenanceTemporarily(StorageNodeInfo nodeInfo, Clust
-        if (!otherGroupCheck.settingWantedStateIsAllowed()) {
-            return otherGroupCheck;
+        if (moreThanOneGroupAllowedToBeDown()) {
+            Result otherGroupCheck = nodesInOtherGroups(nodeInfo.getNode());
```
Yes, good point. Working on rewriting this, some of the methods for checking if a nodes is allowed to be set to maintenance do not fit in too well with allowing more than one node/group to go down at the same time, so need to look more into that.

Commit:
Reimplement checking of other nodes and nodes in groups being not up

Commit:
Merge branch 'master' into hmusum/allow-groups-to-be-down

Commit:
Add back original code and use it for default config

Set default value for max_number_of_groups_allowed_to_be_down to -1
and use that to switch between old and new code

Commit:
Parameterize tests

hmusum commented:
Code for allowing more than one groups to be down should now be functional and only be used when max_number_of_groups_allowed_to_be_down != -1 (-1 is the default value). Config setting, feature flag to control this etc. will come in a later PR. @hakonhall please take a look

hakonhall commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -441,4 +527,14 @@ private Result checkDistributors(Node node, int clusterStateVersion) {
 
+    private Set<Integer> groupsWithStorageNodesWantedStateNotUp() {
+        return clusterInfo.getStorageNodeInfos().stream()
+                          .filter(sni -> !UP.equals(sni.getWantedState().getState()))
```
The original method is `otherNodeInGroupHasWantedState`, which differs from this on 2 accounts:
 - Use `getUserWantedState` instead of `getWantedState`?  As it is there is no practical difference, but getUserWantedState corresponds 1:1 with the state we're about approve/deny with the NodeStateChangeChecker.
 - Also check the distributor?  Only matters if an operator is setting (user) wanted state.

hakonhall commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -268,6 +284,60 @@ private Result anotherNodeInAnotherGroupHasWantedState(StorageNodeInfo nodeInfo)
+        Node node = nodeInfo.getNode();
+
+        if (groupVisiting.isHierarchical()) {
+            if (maxNumberOfGroupsAllowedToBeDown <= 1) {
```
Remove this `if` branch?  Since `maxNumberOfGroupsAllowedToBeDown == -1` is handled elsewhere and serve as the old behavior, we can handle `maxNumberOfGroupsAllowedToBeDown == 1` in the same way as `maxNumberOfGroupsAllowedToBeDown > 1` and by this `else` branch.

hmusum commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -50,6 +56,8 @@ public NodeStateChangeChecker(ContentCluster cluster, boolean inMoratorium) {
         this.clusterInfo = cluster.clusterInfo();
         this.inMoratorium = inMoratorium;
         this.maxNumberOfGroupsAllowedToBeDown = cluster.maxNumberOfGroupsAllowedToBeDown();
+        if ( ! groupVisiting.isHierarchical() && maxNumberOfGroupsAllowedToBeDown > 1)
```
Will be handled when adding code to configure this in services.xml (there is no east way of checking config overrides for a specific config today)

Commit:
Check state down later and simplify

hmusum commented:
Thanks, I think I have addressed all concerns now

hakonhall commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -268,6 +284,37 @@ private Result anotherNodeInAnotherGroupHasWantedState(StorageNodeInfo nodeInfo)
+                return Optional.empty();
+            if (groupsWithStorageNodesWantedStateNotUp.size() < maxNumberOfGroupsAllowedToBeDown)
+                return Optional.of(allowSettingOfWantedState());
+            if (aGroupContainsNode(groupsWithStorageNodesWantedStateNotUp, node))
```
```
            if (aGroupContainsNode(sameGroups, node))
```

hakonhall commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -268,6 +284,37 @@ private Result anotherNodeInAnotherGroupHasWantedState(StorageNodeInfo nodeInfo)
+            String disallowMessage = "At most nodes in " + maxNumberOfGroupsAllowedToBeDown + " groups can have wanted state";
+            if (groupsWithStorageNodesWantedStateNotUp.size() == 0)
+                return Optional.empty();
+            if (groupsWithStorageNodesWantedStateNotUp.size() < maxNumberOfGroupsAllowedToBeDown)
```
```
var sameGroups = ... // groups with at least one node with the same state & description
if (sameGroups.size() == 0) return disallow;
var notUpOrRetiredGroups = ...
if (notUpOrRetiredGroups.size() < maxNumberOfGroupsAllowedToBeDown) allow;
```

Commit:
Handle case where a node has another description for wanted state

Also add group indexes for disallow messages where relevant

hakonhall commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -441,4 +535,41 @@ private Result checkDistributors(Node node, int clusterStateVersion) {
+    }
+
+    // groups with at least one node with the same state & description
+    private Set<Integer> retiredOrNotUpGroups(ClusterState clusterState, State... states) {
```
This should be similar to `checkAllNodesAreUp()`, so find those groups that have at least one distributor or storage node with either a non-UP/RETIRED user wanted state, or its state is != UP/RETIRED.

hmusum replied:
Fixed

Commit:
FIx retiredOrNotUpGroups()

hakonhall commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -441,4 +535,43 @@ private Result checkDistributors(Node node, int clusterStateVersion) {
+    // groups with at least one node with the same state & description
+    private Set<Integer> retiredOrNotUpGroups(ClusterState clusterState) {
+        return clusterInfo.getAllNodeInfos().stream()
+                          .filter(nodeInfo -> (nodeInfo.getUserWantedState().getState() == RETIRED
```
Filter those that have user wanted state *!=* RETIRED *and* != UP, *or* those that have state *!=* RETIRED *and* != UP.  We essentially treat RETIRED as UP in these checks.  If either the user wanted state or the state is not up/retired, we want to add the corresponding group to the returned set.

hmusum replied:
Fixed

Commit:
Rename and fix logic

hmusum commented:
@vekterli please take a look

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -268,6 +287,65 @@ private Result anotherNodeInAnotherGroupHasWantedState(StorageNodeInfo nodeInfo)
+    }
+
+    private ArrayList<Integer> sortSetIntoList(Set<Integer> retiredOrNotUpGroups) {
+        var groupsWithNodesDown = new ArrayList<>(retiredOrNotUpGroups);
```
Should the parameter and variable names be unified here? It could be a bit confusing how `retiredOrNotUpGroups` suddenly becomes `groupsWithNodesDown`.

hmusum replied:
Yes, fixed

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -268,6 +287,65 @@ private Result anotherNodeInAnotherGroupHasWantedState(StorageNodeInfo nodeInfo)
+
+            Set<Integer> retiredAndNotUpGroups = groupsWithNotRetiredAndNotUp(clusterState);
+            int numberOfGroupsToConsider = retiredAndNotUpGroups.size();
+            // Subtract one group if node is in a group with nodes already retired or not up, since number of such groups wil
```
typo nit
```suggestion
            // Subtract one group if node is in a group with nodes already retired or not up, since number of such groups will
```

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -441,4 +535,43 @@ private Result checkDistributors(Node node, int clusterStateVersion) {
+                          .filter(sni -> !UP.equals(sni.getUserWantedState().getState()))
+                          .map(NodeInfo::getGroup)
+                          .filter(Objects::nonNull)
+                          .filter(Group::isLeafGroup)
```
In a valid configuration nodes should always be configured as part of leaf groups, so I think this is always true. But it doesn't hurt to be explicit about the expectations of the node.

vekterli reviewed (dismissed):
:+1:

Commit:
Update clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java

Co-authored-by: Tor Brede Vekterli <vekterli@yahooinc.com>

Commit:
Rename variables
